lua/entities/gmod_wire_expression2/core/wirelink.lua

```diff
@@ -635,3 +635,96 @@ e2function number wirelink:writeTableSimple(address, table data)
 	if not validWirelink(self, this) or not this.WriteCell then return 0 end
 	return writeArraySimple(this, address, data.n)
 end
+
+-- Events for when THIS e2 is read or written to via wirelink, hispeed, whatever uses read & write cell.
+
+-- Events can't have return values, lambdas can't have multiple return types
+-- Don't want to risk a magic number as error causing a collision, so I figure
+-- functions for returning hispeed error & value are the best compromise.
+
+__e2setcost(2)
+e2function void hispeedReturnValue(number value)
+	self.data.hispeedIOError = false
+	self.data.readCellValue = value
+end
+
+e2function void hispeedSetError(number value)
+	self.data.hispeedIOError = value ~= 0
+end
+
+E2Lib.registerEvent("readCell",
+	{
+		{"Address","n"}
+	},
+	function(ctx) -- Constructor
+		local function readCellEvent(self,addr)
+			if self.error then return nil end
+			ctx.data.hispeedIOError = false
+			ctx.data.readCellValue = 0
+			self:ExecuteEvent("readCell",{addr})
+			if ctx.data.hispeedIOError or self.error then return nil end
+			return ctx.data.readCellValue
+		end
+		local ent = ctx.entity
+		if ent.ReadCell == ent.E2ReadCellEvent or not ent.ReadCell then
+			ent.ReadCell = readCellEvent
+			ent.E2ReadCellEvent = readCellEvent
+		else
+			ctx:throw("E2's entity has a different ReadCell function present already! Disable strict to overwrite.")
+			-- Good idea to save so we can restore it after, best not to brick our entity's I/O
+			ent.E2OldReadCell = ent.ReadCell
+			ent.ReadCell = readCellEvent
+		end
+	end,
+	function(ctx)
+		local ent = ctx.entity
+		-- If it's not ours, then whatever context gave us the function can handle its removal.
+		if ent.ReadCell == ent.E2ReadCellEvent then
+			if ent.OldE2ReadCell then
+				ent.ReadCell = ent.OldE2ReadCell
+			else
+				ent.ReadCell = nil
+			end
+		end
+		ent.E2ReadCellEvent = nil
+		ent.OldE2ReadCell = nil
+	end
+)
+
+
+E2Lib.registerEvent("writeCell",
+	{
+		{"Address","n"},
+		{"Value","n"}
+	},
+	function(ctx) -- Constructor
+		local function writeCellEvent(self,addr,value)
+			if self.error then return nil end
+			ctx.data.hispeedIOError = false
+			self:ExecuteEvent("writeCell",{addr,value})
+			if ctx.data.hispeedIOError or self.error then return nil end
+			return true
+		end
+		local ent = ctx.entity
+		if ent.WriteCell == ent.E2WriteCellEvent or not ent.WriteCell then
+			ent.WriteCell = writeCellEvent
+			ent.E2WriteCellEvent = writeCellEvent
+		else
+			ctx:throw("E2's entity has a different WriteCell function present already! Remove @strict to overwrite.")
+			ent.OldE2WriteCell = ent.WriteCell
+			ent.WriteCell = writeCellEvent
+		end
+	end,
+	function(ctx)
+		local ent = ctx.entity
+		if ent.WriteCell == ent.E2WriteCellEvent then
+			if ent.OldE2WriteCell then
+				ent.WriteCell = ent.OldE2WriteCell
+			else
+				ent.WriteCell = nil
+			end
+		end
+		ent.E2WriteCellEvent = nil
+		ent.OldE2WriteCell = nil
+	end
+)
```

Comment:
Can we not define these functions in the E2 init.lua instead of doing this hacky stuff?

Reply:
The part I'm concerned with is that defining readcell & writecell for all [will identify every E2 as a hispeed device with isHiSpeed](https://github.com/wiremod/wire/blob/e163a7d4f9857f04bac50d88a27ae30bc5128093/lua/entities/gmod_wire_expression2/core/wirelink.lua#L145-L148), even E2's that aren't listening to either event.
Most of my other reasons(regarding someone accidentally deleting another entity's read/write by using it as the entity of an e2 context) would be solved by just defining them under the E2 Chip's init.lua I think.

Reply:
I think isHispeed should return 1, so putting it in init.lua is fine.